### DISCUSSION TITLE
Viewport and 2D Camera for zooming and panning on an infinite canvas

### DISCUSSION THREAD

**Discussion by ctwhome** [pixijs/pixijs#10371] [v8 General]
Hallo! Congratulations and thank you for the release of the V8! 

I need some guidance on how to build an infinite canvas with PixiV8.

There are mentions in several places about how V8 is introducing a viewport/2d camera (I see in this [comment](https://github.com/pixijs/pixijs/issues/8809#issuecomment-1304560297) and in the blogpost of the v8 release). I suppose that the introduction of Render Groups is a way to create something like an infinite canvas with zooming and panning the 2D camera (container). Is there any example to start with? Would it also replace the [pixi-viewport plugin demo](https://davidfig.github.io/pixi-viewport/)

![image](https://github.com/pixijs/pixijs/assets/4195550/40c08557-9830-46db-b21f-f44977b5dacf)

Many thanks!

**Comment by ctwhome** [comment]
I made some progress by modifying the new renderGroup example from the docs. 
[playgound link](https://pixijs.com/8.x/playground?state=eyJmaWxlcyI6eyIvc3JjL2luZGV4LmpzIjp7ImNvZGUiOiJpbXBvcnQgeyBBcHBsaWNhdGlvbiwgQXNzZXRzLCBDb250YWluZXIsIFNwcml0ZSB9IGZyb20gJ3BpeGkuanMnO1xuXG4oYXN5bmMgKCkgPT4ge1xuICAgIC8vIENyZWF0ZSBhIG5ldyBhcHBsaWNhdGlvblxuICAgIGNvbnN0IGFwcCA9IG5ldyBBcHBsaWNhdGlvbigpO1xuXG4gICAgLy8gSW5pdGlhbGl6ZSB0aGUgYXBwbGljYXRpb25cbiAgICBhd2FpdCBhcHAuaW5pdCh7XG4gICAgICAgIGJhY2tncm91bmRDb2xvcjogJ2Jyb3duJyxcbiAgICAgICAgcmVzaXplVG86IHdpbmRvdyxcbiAgICB9KTtcblxuICAgIGNvbnN0IHRyZWVUZXh0dXJlID0gYXdhaXQgQXNzZXRzLmxvYWQoYGh0dHBzOi8vcGl4aWpzLmNvbS9hc3NldHMvdHJlZS5wbmdgKTtcblxuICAgIGNvbnN0IHdvcmxkQ29udGFpbmVyID0gbmV3IENvbnRhaW5lcih7XG4gICAgICAgIC8vIHRoaXMgd2lsbCBtYWtlIG1vdmluZyB0aGlzIGNvbnRhaW5lciBHUFUgcG93ZXJlZFxuICAgICAgICBpc1JlbmRlckdyb3VwOiB0cnVlLFxuICAgIH0pO1xuXG4gICAgY29uc3Qgd29ybGRTaXplID0gNTAwMDtcblxuICAgIGZvciAobGV0IGkgPSAwOyBpIDwgMTAwMDA7IGkrKykge1xuICAgICAgICBjb25zdCB5UG9zID0gTWF0aC5yYW5kb20oKSAqIHdvcmxkU2l6ZTtcblxuICAgICAgICBjb25zdCB0cmVlID0gbmV3IFNwcml0ZSh7XG4gICAgICAgICAgICB0ZXh0dXJlOiB0cmVlVGV4dHVyZSxcbiAgICAgICAgICAgIHg6IE1hdGgucmFuZG9tKCkgKiB3b3JsZFNpemUsXG4gICAgICAgICAgICB5OiB5UG9zLFxuICAgICAgICAgICAgc2NhbGU6IDAuMjUsXG4gICAgICAgICAgICBhbmNob3I6IDAuNSxcbiAgICAgICAgfSk7XG5cbiAgICAgICAgd29ybGRDb250YWluZXIuYWRkQ2hpbGQodHJlZSk7XG4gICAgfVxuXG4gICAgLy8gc29ydCB0aGUgdHJlZXMgYnkgdGhlaXIgeSBwb3NpdGlvblxuICAgIHdvcmxkQ29udGFpbmVyLmNoaWxkcmVuLnNvcnQoKGEsIGIpID0%2BIGEucG9zaXRpb24ueSAtIGIucG9zaXRpb24ueSk7XG5cbiAgICBhcHAuc3RhZ2UuYWRkQ2hpbGQod29ybGRDb250YWluZXIpO1xuXG4gICAgLy8gQXBwZW5kIHRoZSBhcHBsaWNhdGlvbiBjYW52YXMgdG8gdGhlIGRvY3VtZW50IGJvZHlcbiAgICBkb2N1bWVudC5ib2R5LmFwcGVuZENoaWxkKGFwcC5jYW52YXMpO1xuXG4gICAgLy8gUGFubmluZyBzdXBwb3J0XG4gICAgbGV0IGlzRHJhZ2dpbmcgPSBmYWxzZTtcbiAgICBsZXQgcHJldmlvdXNQb3NpdGlvbiA9IHsgeDogMCwgeTogMCB9O1xuXG4gICAgYXBwLmNhbnZhcy5hZGRFdmVudExpc3RlbmVyKCdtb3VzZWRvd24nLCAoZSkgPT4ge1xuICAgICAgICBpc0RyYWdnaW5nID0gdHJ1ZTtcbiAgICAgICAgcHJldmlvdXNQb3NpdGlvbi54ID0gZS5jbGllbnRYO1xuICAgICAgICBwcmV2aW91c1Bvc2l0aW9uLnkgPSBlLmNsaWVudFk7XG4gICAgfSk7XG5cbiAgICB3aW5kb3cuYWRkRXZlbnRMaXN0ZW5lcignbW91c2V1cCcsICgpID0%2BIHtcbiAgICAgICAgaXNEcmFnZ2luZyA9IGZhbHNlO1xuICAgIH0pO1xuXG4gICAgYXBwLmNhbnZhcy5hZGRFdmVudExpc3RlbmVyKCdtb3VzZW1vdmUnLCAoZSkgPT4ge1xuICAgICAgICBpZiAoaXNEcmFnZ2luZykge1xuICAgICAgICAgICAgY29uc3QgZHggPSBlLmNsaWVudFggLSBwcmV2aW91c1Bvc2l0aW9uLng7XG4gICAgICAgICAgICBjb25zdCBkeSA9IGUuY2xpZW50WSAtIHByZXZpb3VzUG9zaXRpb24ueTtcblxuICAgICAgICAgICAgd29ybGRDb250YWluZXIueCArPSBkeDtcbiAgICAgICAgICAgIHdvcmxkQ29udGFpbmVyLnkgKz0gZHk7XG5cbiAgICAgICAgICAgIHByZXZpb3VzUG9zaXRpb24ueCA9IGUuY2xpZW50WDtcbiAgICAgICAgICAgIHByZXZpb3VzUG9zaXRpb24ueSA9IGUuY2xpZW50WTtcbiAgICAgICAgfVxuICAgIH0pO1xuXG4gICAgLy8gWm9vbSBzdXBwb3J0IGNlbnRlcmVkIG9uIHRoZSBtb3VzZSBwb3NpdGlvblxuICAgIGFwcC5jYW52YXMuYWRkRXZlbnRMaXN0ZW5lcignd2hlZWwnLCAoZSkgPT4ge1xuICAgICAgICBlLnByZXZlbnREZWZhdWx0KCk7IC8vIFByZXZlbnQgdGhlIHBhZ2UgZnJvbSBzY3JvbGxpbmdcblxuICAgICAgICBjb25zdCBzY2FsZUZhY3RvciA9IGUuZGVsdGFZIDwgMCA%2FIDEuMSA6IDAuOTsgLy8gQWRqdXN0aW5nIHRoZSB6b29tIHNwZWVkXG4gICAgICAgIGNvbnN0IG1vdXNlUG9zaXRpb24gPSB7IHg6IGUuY2xpZW50WCwgeTogZS5jbGllbnRZIH07XG5cbiAgICAgICAgLy8gQ29udmVydCBtb3VzZSBwb3NpdGlvbiB0byB0aGUgd29ybGQgY29udGFpbmVyJ3MgbG9jYWwgc3BhY2VcbiAgICAgICAgY29uc3QgbG9jYWxQb3MgPSB7XG4gICAgICAgICAgICB4OiAobW91c2VQb3NpdGlvbi54IC0gd29ybGRDb250YWluZXIueCkgLyB3b3JsZENvbnRhaW5lci5zY2FsZS54LFxuICAgICAgICAgICAgeTogKG1vdXNlUG9zaXRpb24ueSAtIHdvcmxkQ29udGFpbmVyLnkpIC8gd29ybGRDb250YWluZXIuc2NhbGUueSxcbiAgICAgICAgfTtcblxuICAgICAgICAvLyBBcHBseSBzY2FsZSBmYWN0b3JcbiAgICAgICAgd29ybGRDb250YWluZXIuc2NhbGUueCAqPSBzY2FsZUZhY3RvcjtcbiAgICAgICAgd29ybGRDb250YWluZXIuc2NhbGUueSAqPSBzY2FsZUZhY3RvcjtcblxuICAgICAgICAvLyBBZGp1c3QgY29udGFpbmVyIHBvc2l0aW9uIGJhc2VkIG9uIG5ldyBzY2FsZVxuICAgICAgICB3b3JsZENvbnRhaW5lci54ID0gbW91c2VQb3NpdGlvbi54IC0gKGxvY2FsUG9zLnggKiB3b3JsZENvbnRhaW5lci5zY2FsZS54KTtcbiAgICAgICAgd29ybGRDb250YWluZXIueSA9IG1vdXNlUG9zaXRpb24ueSAtIChsb2NhbFBvcy55ICogd29ybGRDb250YWluZXIuc2NhbGUueSk7XG4gICAgfSk7XG5cbn0pKCk7XG4ifSwiL3NyYy9zdHlsZXMuY3NzIjp7ImNvZGUiOiJib2R5IHsgbWFyZ2luOiAwOyBvdmVyZmxvdzogaGlkZGVuOyB9In0sIi9pbmRleC5odG1sIjp7ImNvZGUiOiJcbjwhRE9DVFlQRSBodG1sPlxuICA8aHRtbD5cbiAgPGhlYWQ%2BXG4gIDx0aXRsZT5QaXhpSlMgUGxheWdyb3VuZDwvdGl0bGU%2BXG4gIDxtZXRhIGNoYXJzZXQ9XCJVVEYtOFwiIC8%2BXG4gIDxsaW5rIHJlbD1cInN0eWxlc2hlZXRcIiBocmVmPVwic3JjL3N0eWxlcy5jc3NcIj5cbjwvaGVhZD5cbjxib2R5PlxuICA8c2NyaXB0IHNyYz1cInNyYy9pbmRleC5qc1wiPjwvc2NyaXB0PlxuPC9ib2R5PlxuPC9odG1sPiJ9LCIvcGFja2FnZS5qc29uIjp7ImNvZGUiOiJ7XG4gIFwic2NyaXB0c1wiOiB7XG4gICAgXCJzdGFydFwiOiBcInBhcmNlbCBpbmRleC5odG1sIC0tb3BlblwiLFxuICAgIFwiYnVpbGRcIjogXCJwYXJjZWwgYnVpbGQgaW5kZXguaHRtbFwiXG4gIH0sXG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcInBpeGkuanNcIjogXCI4LjAuMlwiXG4gIH0sXG4gIFwiZGV2RGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcInBhcmNlbC1idW5kbGVyXCI6IFwiXjEuNi4xXCIsXG4gICAgXCJAYmFiZWwvY29yZVwiOiBcIl43LjIxLjNcIixcbiAgICBcIkBiYWJlbC9wbHVnaW4tcHJvcG9zYWwtY2xhc3MtcHJvcGVydGllc1wiOiBcIl43LjEwLjFcIlxuICB9LFxuICBcIm1haW5cIjogXCIvaW5kZXguaHRtbFwiXG59In0sIi8uYmFiZWxyYyI6eyJjb2RlIjoie1xuICBcInByZXNldHNcIjogW1xuICAgIFtcbiAgICAgIFwiQGJhYmVsL3ByZXNldC1lbnZcIixcbiAgICAgIHtcbiAgICAgICAgXCJ1c2VCdWlsdEluc1wiOiBcImVudHJ5XCIsXG4gICAgICAgIFwiY29yZWpzXCI6IFwiMy4yOVwiLFxuICAgICAgICBcInRhcmdldHNcIjogXCJsYXN0IDIgY2hyb21lIHZlcnNpb25zXCJcbiAgICAgIH1cbiAgICBdXG4gIF0sXG4gIFwicGx1Z2luc1wiOiBbXG4gICAgXCJAYmFiZWwvcGx1Z2luLXByb3Bvc2FsLWNsYXNzLXByb3BlcnRpZXNcIlxuICBdXG59In0sIi9zYW5kYm94LmNvbmZpZy5qc29uIjp7ImNvZGUiOiJ7XG4gICAgICAgICAgICAgICAgXCJpbmZpbml0ZUxvb3BQcm90ZWN0aW9uXCI6IGZhbHNlXG4gICAgICAgICAgICB9In19LCJ2aXNpYmxlRmlsZXMiOlsiL3NyYy9pbmRleC5qcyJdLCJhY3RpdmVGaWxlIjoiL3NyYy9pbmRleC5qcyJ9)

![pan and zoom](https://github.com/pixijs/pixijs/assets/4195550/ee646751-40f3-46c1-90a4-3cfbd3b235ad)

**Reply by papa3500** [reply]
The Playground link has been deleted. Is it possible to restore it? I regret not backing up a month ago.

**Reply by IsmailKefokeris** [reply]
Could you re-link that playground link please?

**Reply by ctwhome** [reply]
Lost it too, pixi needs to fix the playground for this file I guess

**Comment by reececomo** [comment]
As an end-user I agree there is still a teensy bit of ambiguity here.

In say `pixi-viewport` - would you make the viewport container `isRenderGroup=true`, or would you add the contents of the camera to a container, and make that `isRenderGroup=true`?

i.e.

```ts
const viewport = new Viewport({ /* ... */ });
const worldStage = new Container({ isRenderGroup: true });

viewport.addChild(worldStage);

worldStage.addChild(
  sprite1,
  sprite2,
  sprite3,
  // ...
);
```

or

```ts
const viewport = new Viewport({ /* ... */ });
viewport.isRenderGroup = true;

const worldStage = new Container();
viewport.addChild(worldStage);

worldStage.addChild(
  sprite1,
  sprite2,
  sprite3,
  // ...
);
```
